python-package/lightgbm/sklearn.py

```diff
@@ -151,14 +151,18 @@ def __call__(self, preds: np.ndarray, dataset: Dataset) -> Tuple[np.ndarray, np.
             The value of the second order derivative (Hessian) of the loss
             with respect to the elements of preds for each sample point.
         """
-        labels = dataset.get_label()
+        labels = dataset.get_field("label")
```

Comment:
After correcting the type hints in `Dataset` methods, `mypy` rightly started complaining about these parts of the `sklearn` interface, where custom objective functions and metric functions are called.

With many instances like this:

```text
sklearn.py:157: error: Argument 1 has incompatible type "Union[List[float], List[int], ndarray[Any, Any], Any, None]"; expected "Optional[ndarray[Any, Any]]"  [arg-type]
```

That's because we say in documentation and type hints that `sklearn` custom objective functions and eval metric functions should only expect to be passed `numpy` arrays or `None`.

https://github.com/microsoft/LightGBM/blob/7c9a985a476309ba1d7a7767b5247a3d4eb579b9/python-package/lightgbm/sklearn.py#L35-L51

https://github.com/microsoft/LightGBM/blob/7c9a985a476309ba1d7a7767b5247a3d4eb579b9/python-package/lightgbm/sklearn.py#L101-L115

This PR proposes fixing that by switching from `get_label()` to `get_field("label")` (and similar for all the other metadata stuff). Custom metrics and eval functions in the `sklearn` interface should only ever be called on already-constructed `Dataset` objects, so that should be safe. It should also reduce the risk of soem mistake accidentally leading to something other than numpy arrays being passed into `sklearn` custom objective and metric functions.

Reply:
Have you looked into the performance impact of this? This method is called on each iteration and we're currently just returning an attribute from the dataset but this would require making a call to C++ to get the field each time. I think it just returns the pointer so it may not be that expensive but it's something to consider given that we'd be making several calls to get all the fields that this method needs.

Reply:
No, I didn't consider that this would be a performance issue. You're right, it could be!

> I think it just returns the pointer

Nah, your instincts that this might be more expensive were good! This call to `LGBM_DatasetGetField()` ...

https://github.com/microsoft/LightGBM/blob/63a882b7400a19485727cc808e4715897e63635a/python-package/lightgbm/basic.py#L2522-L2527

... is just populating a pointer ...

https://github.com/microsoft/LightGBM/blob/63a882b7400a19485727cc808e4715897e63635a/src/c_api.cpp#L1697

https://github.com/microsoft/LightGBM/blob/63a882b7400a19485727cc808e4715897e63635a/src/io/dataset.cpp#L956

...but then that get materialized into a new `numpy` array (e.g. a full copy is made in memory):

https://github.com/microsoft/LightGBM/blob/63a882b7400a19485727cc808e4715897e63635a/python-package/lightgbm/basic.py#L2532-L2537

https://github.com/microsoft/LightGBM/blob/63a882b7400a19485727cc808e4715897e63635a/python-package/lightgbm/basic.py#L374

So I think you're right, calling `get_field()` unambiguously might not be a good way to do this. That's just wasted effort relative to `.get_label()`, `.get_weight()`, etc., since, on a constructed `Dataset`, `.label`, `.weight`, etc. should already contain `numpy` arrays regardless of what format the raw data was passed in.

Thanks to these calls in `Dataset._lazy_init()`:

https://github.com/microsoft/LightGBM/blob/63a882b7400a19485727cc808e4715897e63635a/python-package/lightgbm/basic.py#L1958-L1967

So given all that....let me think about whether there's a better way to resolve these errors from `mypy`. I don't think they're just things we should ignore... I do feel we should add some stronger guarantees here that by the time custom objective and metric functions should only be passed an already-constructed `Dataset`.

Thanks for bringing it up!

Reply:
Alright @jmoralez I think I found a better way to do this. I just pushed https://github.com/microsoft/LightGBM/pull/6108/commits/017e5e5703bf14f5828fa74d4854b4c72a91f953, with the following changes:

* switches back to using `get_label()` and `get_weight()` instead of `get_field()`
    - *because those will return the data cached on the `Dataset` object without reconstructing a new `numpy` array on every iteration*
* keeps `get_field("group")` for `group`
    - *that'll still introduce a performance penalty (reconstructing an array on every iteration if the objective function / eval metric takes 4 arguments, i.e. is for learning-to-rank)*
    - *but I think we need to do that... right now the docs say such functions should expect `group` to be a `numpy` array, but `lightgbm` is currently passing them a `list`*
    - https://github.com/microsoft/LightGBM/blob/d45dca702eed0d0c8e79c2ca0085096465756ff5/python-package/lightgbm/sklearn.py#L115
* fixes the `mypy` errors by putting `get_label()` and `get_weight()` calls inside functions that use `assert isinstance(x, np.ndarray)` for type narrowing*
* adds docs to `Dataset.get_{group/init_score/weight/position/label}` explaining that the return type will be different for a constructed Dataset

@jmoralez whenever you have time, could you please take another look? Thanks again for bringing this up!

Reply:
Can't we do for group what we do for other fields like init_score and weight where after we set the field we set the Dataset attribute to the result of get_field?
https://github.com/microsoft/LightGBM/blob/8ed371cee49cf86740b25dd9a4b985a75c9f2dba/python-package/lightgbm/basic.py#L2742-L2743
https://github.com/microsoft/LightGBM/blob/8ed371cee49cf86740b25dd9a4b985a75c9f2dba/python-package/lightgbm/basic.py#L2720-L2721

Reply:
We could but that would be a user-facing breaking change. Look at the test cases in this PR... even after construction, `Dataset.get_group()` returns whatever was passed in (most commonly a list).

Do you think it's worth the breaking change to get that consistency?

Reply:
I think it is, it's weird having all of the arguments of a custom objective be arrays except for group. Although the main difference I think is not really the data structure but the format that they're in (group boundaries vs lengths). Given that the docstring says array and it's more consistent with the other fields I'd prefer we override it with get_field. On the other hand I haven't used ranking a lot so I'm not sure which format is more convenient for the objective function.

Reply:
Ok I think your reasoning makes sense and that we should change it to an array.

I'm especially since tonight I found that it'll be set to an array of boundaries the first time you call `get_group()` on a `Dataset` where the `.group` attribute is `None`, which can happen when loading from a binary file:

```python
import lightgbm as lgb
import numpy as np

X = np.array([[1.0, 2.0], [3.0, 4.0]])

dtrain = lgb.Dataset(
    X,
    params={
        "min_data_in_bin": 1,
        "min_data_in_leaf": 1,
        "verbosity": -1
    },
    group=[1, 1],
    label=[1, 2],
)
dtrain.construct()

# get_group() returns a list of sizes
assert dtrain.get_group() == [1, 1]

# get_field() returns a numpy array of bounadries
np.testing.assert_array_equal(
    dtrain.get_field("group"),
    np.array([0, 1, 2])
)


# round-trip to and from binary file
dtrain.save_binary("dtrain.bin")
dtrain2 = lgb.Dataset(
    data="dtrain.bin",
    params={
        "min_data_in_bin": 1,
        "min_data_in_leaf": 1,
        "verbosity": -1
    }
)

# before construction, group is empty
assert dtrain2.group is None

# after construction, get_group() returns a numpy array of sizes
dtrain2.construct()
np.testing.assert_array_equal(
    dtrain.get_group(),
    np.array([1, 1])
)

# ... and get_field() returns a numpy array of boundaries
np.testing.assert_array_equal(
    dtrain.get_field("group"),
    np.array([0, 1, 2])
)
```

That doesn't matter specifically for the `scikit-learn` interface (which I don't believe could ever encounter a binary dataset file), but it does give me more confidence that other parts of the code base aren't implicitly relying on `Dataset.group` being a list.

----

I just did the following:

* pushed https://github.com/microsoft/LightGBM/pull/6108/commits/ada18f80441262ce3a191c0c55993e0b37ff3f68 with these changes
* changed the label on this PR to `breaking`
* changed the title of the PR
* added a section in the description explaining why this is `breaking`

Thanks for talking through it with me, I know this is way down in the depths of the library and that reviewing it takes a lot of effort.

Reply:
I won't merge this until you've had another chance to review @jmoralez . Take your time!